Source/Core/Core/IOS/USB/Emulated/Skylanders/SkylanderFigure.cpp

```diff
@@ -0,0 +1,455 @@
+{
+  // Set the sector permissions
+  u32 first_block = 0x690F0F0F;
+  u32 other_blocks = 0x69080F7F;
+  memcpy(&m_data[0x36], &first_block, sizeof(first_block));
+  for (size_t index = 1; index < 0x10; index++)
+  {
+    memcpy(&m_data[(index * 0x40) + 0x36], &other_blocks, sizeof(other_blocks));
+  }
+}
+
+void SkylanderFigure::PopulateKeys()
+{
+  for (u8 sector = 0; sector < 0x10; sector++)
+  {
+    u16 key_offset = (sector * 64) + (3 * 16);
+    u64 key = CalculateKeyA(sector, std::span<u8, 4>(m_data.begin(), 4));
+
+    for (u32 j = 0; j < 6; j++)
+    {
+      u16 index = key_offset + (5 - j);
+      u8 byte = (key >> (j * 8)) & 0xFF;
+      m_data[index] = byte;
+    }
+  }
+}
+
+SkylanderFigure::SkylanderFigure(const std::string& file_path)
+{
+  m_sky_file = File::IOFile(file_path, "w+b");
+  m_data = {};
+}
+// Generate a AES key without the block filled in
+void SkylanderFigure::GenerateIncompleteHashIn(u8* dest) const
+{
+  std::array<u8, 0x56> hash_in = {};
+
+  // copy first 2 blocks into hash
+  GetBlock(0, hash_in.data());
+  GetBlock(1, hash_in.data() + 0x10);
+
+  // Skip 1 byte. Is a block index that needs to be set per block.
+
+  // Byte array of ascii string " Copyright (C) 2010 Activision. All Rights Reserved.". The space at
+  // the start of the string is intentional
+  static constexpr std::array<u8, 0x35> HASH_CONST = {
+      0x20, 0x43, 0x6F, 0x70, 0x79, 0x72, 0x69, 0x67, 0x68, 0x74, 0x20, 0x28, 0x43, 0x29,
+      0x20, 0x32, 0x30, 0x31, 0x30, 0x20, 0x41, 0x63, 0x74, 0x69, 0x76, 0x69, 0x73, 0x69,
+      0x6F, 0x6E, 0x2E, 0x20, 0x41, 0x6C, 0x6C, 0x20, 0x52, 0x69, 0x67, 0x68, 0x74, 0x73,
+      0x20, 0x52, 0x65, 0x73, 0x65, 0x72, 0x76, 0x65, 0x64, 0x2E, 0x20};
+
+  memcpy(hash_in.data() + 0x21, HASH_CONST.data(), HASH_CONST.size());
+
+  memcpy(dest, hash_in.data(), 0x56);
+}
+void SkylanderFigure::Encrypt(std::span<u8, FIGURE_SIZE> input)
+{
+  std::array<u8, 0x56> hash_in = {};
+
+  GenerateIncompleteHashIn(hash_in.data());
+
+  std::array<u8, FIGURE_SIZE> encrypted = {};
+
+  std::array<u8, 0x10> current_block = {};
+
+  // Run for every block
+  for (u8 i = 0; i < 64; ++i)
+  {
+    memcpy(current_block.data(), input.data() + (i * 0x10), 0x10);
+
+    // Skip sector trailer and the first 8 blocks
+    if (((i + 1) % 4 == 0) || i < 8)
+    {
+      memcpy(encrypted.data() + (i * 0x10), current_block.data(), 0x10);
+      continue;
+    }
+
+    // Block index
+    hash_in[0x20] = i;
+
+    std::array<u8, 0x10> hash_out = {};
+
+    mbedtls_md5_ret(hash_in.data(), 0x56, hash_out.data());
+
+    mbedtls_aes_context aes_context = {};
+
+    mbedtls_aes_setkey_enc(&aes_context, hash_out.data(), 128);
+
+    mbedtls_aes_crypt_ecb(&aes_context, MBEDTLS_AES_ENCRYPT, current_block.data(),
+                          encrypted.data() + (i * 0x10));
+  }
+
+  memcpy(m_data.data(), encrypted.data(), FIGURE_SIZE);
+
+  NOTICE_LOG_FMT(IOS_USB, "Encrypted skylander data: \n{}", HexDump(encrypted.data(), FIGURE_SIZE));
```

Comment:
This should probably be a `DEBUG_LOG_FMT` (our lowest level), not a `NOTICE_LOG_FMT` (our highest level).